controlplane/kubeadm/internal/controllers/controller_test.go

```diff
@@ -1227,88 +1257,460 @@ kubernetesVersion: metav1.16.1`,
+	g.Expect(env.Create(ctx, existingKubeadmConfig, client.FieldOwner("manager"))).To(Succeed())
+
+	// Existing Machine
+	fd := pointer.String("fd1")
+	inPlaceMutatingMachine := &clusterv1.Machine{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Machine",
+			APIVersion: clusterv1.GroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "existing-machine",
+			Namespace: namespace.Name,
+			Labels: map[string]string{
+				"preserved-label": "preserved-value",
+				"dropped-label":   "dropped-value",
+				"modified-label":  "modified-value",
+			},
+			Annotations: map[string]string{
+				"preserved-annotation": "preserved-value",
+				"dropped-annotation":   "dropped-value",
+				"modified-annotation":  "modified-value",
+			},
+		},
+		Spec: clusterv1.MachineSpec{
+			ClusterName: testCluster.Name,
+			Bootstrap: clusterv1.Bootstrap{
+				ConfigRef: bootstrapRef,
+			},
+			InfrastructureRef:       *infraMachineRef,
+			Version:                 pointer.String("v1.25.3"),
+			FailureDomain:           fd,
+			NodeDrainTimeout:        duration5s,
+			NodeVolumeDetachTimeout: duration5s,
+			NodeDeletionTimeout:     duration5s,
+		},
+	}
+	// Note: use "manager" as the field owner to mimic the manager used before ClusterAPI v1.4.0.
+	g.Expect(env.Create(ctx, inPlaceMutatingMachine, client.FieldOwner("manager"))).To(Succeed())
+
+	deletingMachine := &clusterv1.Machine{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: clusterv1.GroupVersion.String(),
+			Kind:       "Machine",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        "deleting-machine",
+			Namespace:   namespace.Name,
+			Labels:      map[string]string{},
+			Annotations: map[string]string{},
+			Finalizers:  []string{"testing-finalizer"},
+		},
+		Spec: clusterv1.MachineSpec{
+			ClusterName: testCluster.Name,
+			InfrastructureRef: corev1.ObjectReference{
+				Namespace: namespace.Name,
+			},
+			Bootstrap: clusterv1.Bootstrap{
+				DataSecretName: pointer.String("machine-bootstrap-secret"),
+			},
+		},
+	}
+	g.Expect(env.Create(ctx, deletingMachine, client.FieldOwner(classicManager))).To(Succeed())
+	// Delete the machine to put it in the deleting state
+	g.Expect(env.Delete(ctx, deletingMachine)).To(Succeed())
+	// Wait till the machine is marked for deletion
+	g.Eventually(func() bool {
+		if err := env.Get(ctx, client.ObjectKeyFromObject(deletingMachine), deletingMachine); err != nil {
+			return false
+		}
+		return !deletingMachine.DeletionTimestamp.IsZero()
+	}, 30*time.Second).Should(BeTrue())
+
+	kcp := &controlplanev1.KubeadmControlPlane{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "KubeadmControlPlane",
+			APIVersion: controlplanev1.GroupVersion.String(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			UID:       types.UID("abc-123-control-plane"),
+			Name:      "existing-kcp",
+			Namespace: namespace.Name,
+		},
+		Spec: controlplanev1.KubeadmControlPlaneSpec{
+			Version:           "v1.25.3",
```

Comment:
nit: would it be possible to use a different version here? Just to also verify that this field is not in-place mutated.

(please also check other non-in-place-mutable fields, should be at least failureDomain and providerID)

Reply:
Added some more checks to make sure that other parts of the spec are not updated.

Reply:
Thx!